src/ImageSearchModal.jsx

```diff
@@ -0,0 +1,134 @@
+import React from 'react';
+import { Button, ListGroup, ListGroupItem, Modal } from 'patternfly-react';
+
+import { ErrorNotification } from './Notification.jsx';
+import * as utils from './util.js';
+import varlink from './varlink.js';
+import cockpit from 'cockpit';
+
+import '../lib/form-layout.less';
+import './ImageSearchModal.css';
+
+const _ = cockpit.gettext;
+
+export class ImageSearchModal extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            imageIdentifier: '',
+            imageList: [],
+            searchInProgress: false,
+        };
+        this.onDownloadClicked = this.onDownloadClicked.bind(this);
+        this.onItemSelected = this.onItemSelected.bind(this);
+        this.onSearchTriggered = this.onSearchTriggered.bind(this);
+        this.onValueChanged = this.onValueChanged.bind(this);
+    }
+
+    componentDidMount() {
+        this._isMounted = true;
+    }
+
+    componentWillUnmount() {
+        this._isMounted = false;
+    }
+
+    onDownloadClicked() {
+        let selectedImageName = this.state.imageList[this.state.selected].name;
+
+        this.props.close();
+        this.props.downloadImage(selectedImageName);
+    }
+
+    onItemSelected(key) {
+        this.setState({ selected: key });
+    }
+
+    onSearchTriggered() {
+        if (!this._isMounted)
+            return;
+
+        varlink.call(utils.PODMAN_ADDRESS, "io.podman.SearchImage", { name: this.state.imageIdentifier })
+                .then(reply => this.setState({ imageList: reply.images || [], searchInProgress: false }))
+                .catch(ex => {
+                    this.setState({
+                        searchInProgress: false,
+                        dialogError: _("Failed to search for new images"),
+                        dialogErrorDetail: cockpit.format("$0: $1", ex.error, ex.parameters && ex.parameters.reason)
+                    });
+                });
+    }
+
+    onValueChanged(key, value) {
+        this.setState({ [key]: value });
+
+        if (key == 'imageIdentifier') {
+            this.setState({ searchInProgress: true });
+            // Clears the previously set timer.
+            clearTimeout(this.typingTimeout);
+
+            // Reset the timer, to make the http call after 1000MS
+            this.typingTimeout = setTimeout(this.onSearchTriggered, 1000);
```

Comment:
Not sure what we usually do here, but 1s seems too much. Maybe 500ms? (Unless we decide for an explicit `Return`)/

@garrett Is there consensus on a good timeout for being sure that text entry has stopped? We're triggering a search here.

Reply:
Perhaps not consensus, but precedent. The [realmd join dialog](https://github.com/cockpit-project/cockpit/blob/master/pkg/realmd/operation.js#L69) uses 1s, like this PR. The [docker UI search dialog](https://github.com/cockpit-project/cockpit/blob/master/pkg/docker/search.js#L78) uses 2 s even.

Reply:
Ah, indeed. Thanks. I guess doing the same here is fine, then (if hitting `Return` works as well).

Reply:
1. Are we querying a remote server?
   - I assume the Docker hub? If so, I looked for stats.
2. How long does a request from the server take to complete a round trip (contact, process, results)?
   - Doing an autocomplete search on hub.docker.com for "fedora" took 253 ms & 251 ms, then 128 & 130 ms (it seems variable between 130 - 250 ms). There are two requests at the same time, which probably affects timing. One search filters by "store,official" and the other by "community". Both of the autocomplete searches limit by 4.
   - When performing a full search, it took 145 ms, which included 25 results with metadata.
   - Autocomplete and a full search both use the same API, but with different limits (`image_filter` and `page_size` for limiting auto-complete and `type=image` with `page_size` as 25 for the full search).
   - hub.docker.com is hosted on us-east-1 on AWS, so it is faster in North America; the stats above are from a (not fantastic) residential cable modem line in Europe, so the timing is probably more on the conservative side
3. Is the server rate limited?
   - Seems like it is not: https://stackoverflow.com/questions/46424132/docker-pull-request-limits-for-docker-hub-registry

We will want to limit to > 2 characters (unless enter is pressed), debounce, throttle, strip trailing spaces, [XMLHttpRequest.abort()](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/abort) unneeded pending requests, and cache (based on stripped query string) to improve the user experience.

_Interestingly enough_, 250ms is the rough median reaction time of humans (oft-cited in UX literature and [backed up by a game](https://www.humanbenchmark.com/tests/reactiontime/statistics) when you subtract the 30 ms screen delay) and usually the upper bound to use for making interfaces feel responsive. As the response is typically 130 but can get up to 250 (at least from my home network) with multiple calls at the same time, we probably want to make 1 call if we can, assume it will return within 150ms, and pad to 250ms (that is 100ms of delay) if we want it to feel _instant_. Search results seem to weight official images at the top and have metadata included about official-ness, so we can probably just do 1 call.

_However_, with a spinner, we can bump up the wait time a little. Still, if it is >= 1 second, people will think it's slow. A medium ground could be to aim for 400ms, perhaps. We should implement something, check metrics and how it feels, and adjust accordingly.

As average typing time is roughly 40 words per minute, 40/60 == ⅔ sec. And people will be thinking as they type, so they'll be a little slower. That's how I wound up with 400 ms as a goal, as a compromise between typing speed and responsiveness.

---

This all said: 1 second is definitely too long, however. 2 is glacial. (The Docker page always felt really slow when searching.) So let's try for 400 ms total time (that is: 250 ms debounce + 150 ms query time) on a first pass, with reasoning above and adjust as needed (if there are "too many" queries or if it feels slow).

It's probably better to start with less UI latency and adjust to add more than the other way around.

Reply:
BTW: Local registries should have a much smaller delay than Docker hub.

Reply:
I honestly don't know how/if it's possible to use  [XMLHttpRequest.abort()](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/abort) to abort unneeded pending requests.
So, to sum up changes for now,
I'll limit the search for >2 chars expect if Enter is hit.
I will strip the input string.
I 'll set the trigger wait time to 400ms instead of 1000ms. But if aborting the ongoing searches is not possible, this will deteriorate performance.
Regarding the caching I also am not sure what you mean. If you mean just store it locally in some dict, I honestly doubt this is a good idea but I 'll leave it up to @martinpitt  and @larskarlitski